renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
+  }
+  else if(opcode == OPCODE_GATHER4_C || opcode == OPCODE_GATHER4_PO_C)
+  {
+    // comparison value
+    StringFormat::snprintf(buf, 255, ", %.10f", lodOrCompareValue);
+
+    sampleProgram = textureDecl + " : register(t" + std::to_string(texSlot) + ");\n" + samplerDecl +
+                    " : register(s" + std::to_string(sampSlot) + ");\n\n";
+    sampleProgram += funcRet + " main() : SV_Target0\n{\nreturn ";
+    sampleProgram += "t.GatherCmp" + strGatherChannel + "(s, " + texcoords + buf + offsets + ")" +
+                     strSwizzle + ";";
+    sampleProgram += "\n}\n";
+  }
+
+  // Create VS/PS to fetch the sample
+  ID3DBlob *vsBlob = NULL;
+  ID3DBlob *psBlob = NULL;
+  UINT flags = D3DCOMPILE_DEBUG | D3DCOMPILE_WARNINGS_ARE_ERRORS | D3DCOMPILE_DEBUG_NAME_FOR_SOURCE;
+  if(m_pDevice->GetShaderCache()->GetShaderBlob(vsProgram.c_str(), "main", flags, "vs_5_0",
+                                                &vsBlob) != "")
+  {
+    RDCERR("Failed to create shader to extract inputs");
+    return false;
+  }
+  if(m_pDevice->GetShaderCache()->GetShaderBlob(sampleProgram.c_str(), "main", flags, "ps_5_0",
+                                                &psBlob) != "")
+  {
+    RDCERR("Failed to create shader to extract inputs");
+    SAFE_RELEASE(vsBlob);
+    return false;
+  }
+
+  // Create a PSO with our VS/PS and all other state from the original event
+  D3D12RenderState &rs = m_pDevice->GetQueue()->GetCommandData()->m_RenderState;
+  D3D12RenderState prevState = rs;
+  WrappedID3D12PipelineState *origPSO =
+      m_pDevice->GetResourceManager()->GetCurrentAs<WrappedID3D12PipelineState>(rs.pipe);
+
+  D3D12_EXPANDED_PIPELINE_STATE_STREAM_DESC pipeDesc;
+  origPSO->Fill(pipeDesc);
+  pipeDesc.VS.BytecodeLength = vsBlob->GetBufferSize();
+  pipeDesc.VS.pShaderBytecode = vsBlob->GetBufferPointer();
+  pipeDesc.PS.BytecodeLength = psBlob->GetBufferSize();
+  pipeDesc.PS.pShaderBytecode = psBlob->GetBufferPointer();
+
+  pipeDesc.GS.BytecodeLength = 0;
+  pipeDesc.GS.pShaderBytecode = NULL;
+  pipeDesc.DS.BytecodeLength = 0;
+  pipeDesc.DS.pShaderBytecode = NULL;
+  pipeDesc.HS.BytecodeLength = 0;
+  pipeDesc.HS.pShaderBytecode = NULL;
+  pipeDesc.CS.BytecodeLength = 0;
+  pipeDesc.CS.pShaderBytecode = NULL;
+
+  ZeroMemory(&pipeDesc.RasterizerState, sizeof(D3D12_RASTERIZER_DESC));
+  pipeDesc.RasterizerState.FillMode = D3D12_FILL_MODE_SOLID;
+  pipeDesc.RasterizerState.CullMode = D3D12_CULL_MODE_NONE;
+  pipeDesc.RasterizerState.FrontCounterClockwise = TRUE;
+
+  pipeDesc.SampleDesc.Count = 1;
+  pipeDesc.SampleDesc.Quality = 0;
+
+  pipeDesc.InputLayout.NumElements = 0;
+  pipeDesc.InputLayout.pInputElementDescs = NULL;
+  pipeDesc.IBStripCutValue = D3D12_INDEX_BUFFER_STRIP_CUT_VALUE_DISABLED;
+  pipeDesc.PrimitiveTopologyType = D3D12_PRIMITIVE_TOPOLOGY_TYPE_TRIANGLE;
+  pipeDesc.RTVFormats.NumRenderTargets = 1;
+  pipeDesc.RTVFormats.RTFormats[0] = retFmt;
+  pipeDesc.DSVFormat = DXGI_FORMAT_UNKNOWN;
```

Comment:
I think there's a few more states to patch - we'd need to disable blending, ensure the SampleMask is correct, sanitise depth and stencil testing, the new view instancing too might break.

Thinking about it, is it easier to only pull in what we actually want? do we need anything from the original PSO apart from the root signature?

Reply:
Yeah, just the root signature. Switched to just creating the PSO from scratch.